src/steamship/agents/examples/document_qa_agent.py

```diff
@@ -0,0 +1,97 @@
+
+        /learn_text
+        { text }
+
+    - An unauthenticated endpoint for answering questions about what it has learned
+
+    This agent provides a starter project for special purpose QA agents that can answer questions about documents
+    you provide.
+    """
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+        # This Mixin provides HTTP endpoints that coordinate the learning of documents.
+        #
+        # It adds the `/learn_url` endpoint which will:
+        #    1) Download the provided URL (PDF, YouTube URL, etc)
+        #    2) Convert that URL into text
+        #    3) Store the text in a vector index
+        #
+        # That vector index is then available to the question answering tool, below.
+        self.add_mixin(IndexerPipelineMixin(self.client, self))
+
+        # A ReACTAgent is an agent that is able to:
+        #    1) Converse with you, casually... but also
+        #    2) Use tools that have been provided to it, such as QA tools or Image Generation tools
+        #
+        # This particular ReACTAgent has been provided with a single tool which will be used whenever
+        # the user answers a question. But you can extend this with more tools if you wish. For example,
+        # you could add tools to generate images, or search Google, or register an account.
+        self._agent = ReACTAgent(
+            tools=[
+                VectorSearchQATool(),  # Tool to answer questions based on a vector store.
+            ],
+            llm=OpenAI(self.client),
+        )
+
+    @post("prompt")
+    def prompt(self, prompt: str) -> str:
+        """Run an agent with the provided text as the input."""
+
+        # AgentContexts serve to allow the AgentService to run agents
+        # with appropriate information about the desired tasking.
+        # Here, we create a new context on each prompt, and append the
+        # prompt to the message history stored in the context.
+        context_id = uuid.uuid4()
+        context = AgentContext.get_or_create(self.client, {"id": f"{context_id}"})
+        context.chat_history.append_user_message(prompt)
+
+        # AgentServices provide an emit function hook to access the output of running
+        # agents and tools. The emit functions fire at after the supplied agent emits
+        # a "FinishAction".
+        #
+        # Here, we show one way of accessing the output in a synchronous fashion. An
+        # alternative way would be to access the final Action in the `context.completed_steps`
+        # after the call to `run_agent()`.
+        output = ""
+
+        def sync_emit(blocks: List[Block], meta: Metadata):
+            nonlocal output
+            block_text = "\n".join(
+                [b.text if b.is_text() else f"({b.mime_type}: {b.id})" for b in blocks]
+            )
+            output += block_text
+
+        context.emit_funcs.append(sync_emit)
+        self.run_agent(self._agent, context)
+        return output
```

Comment:
Yeah I think this might be a good idea; would definitely clean things up in example code -- and still be overridable from the AgentService one is writing, which is one of the properties that @douglas-reid was trying to build in with it (e.g. if you want your REPL harness to include a particular temperature of LLM, or a particular choice of LLM)